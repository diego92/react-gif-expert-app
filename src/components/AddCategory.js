import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
    console.log("handleInputChange llamado");
  };

  const handleSubmit = (e) => {
    e.preventDefault(); // prevenimos el envÃ­o del formulario

    setCategories((cats) => [inputValue, ...cats]);
    setInputValue("");
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange}
      ></input>
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
