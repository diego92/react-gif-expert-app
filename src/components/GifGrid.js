import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
  const [images, setImages] = useState([]);
  useEffect(() => {
    getGifs();
  }, []); // no poner dependecias aca hace que no se ejecute de nuevo la peticion

  const getGifs = async () => {
    const url =
      "https://api.giphy.com/v1/gifs/search?q=dragon+ball&limit=10&api_key=d1DvYQ5f6O59f814O2FX5N3Gca3dBXwX";

    const respuesta = await fetch(url);
    const { data } = await respuesta.json();
    const gifs = data.map((img) => ({
      id: img.id,
      title: img.title,
      url: img.images?.downsized_medium.url,
    }));
    setImages(gifs);
  };

  return (
    <div>
      <h3>{category}</h3>
      {images.map((img) => (
        <GifGridItem key={img.id} {...img} />
      ))}
    </div>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string,
};
