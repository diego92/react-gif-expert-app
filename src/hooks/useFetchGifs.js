import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
  const [state, setstate] = useState({ data: [], loading: true });

  useEffect(() => {
    getGifs(category).then((imgs) => setstate({ data: imgs, loading: false }));
  }, [category]); // si la categoria cambia va a volver a ejecutar el useEffect

  return state; // { data: [], loading: true }
};
