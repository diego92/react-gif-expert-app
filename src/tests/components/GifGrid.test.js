import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe("Pruebas en <GifGrid />", () => {
  const category = "One Punch";

  //   let wrapper = shallow(<GifGrid category={category} />); // no es buena practica pero sino no aplica las ayudas

  //   beforeEach(() => {});

  test("debe mostrar <GifGrid /> correctamente", () => {
    useFetchGifs.mockReturnValue({ data: [], loading: true });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar items cuando se cargan imagenes useFetchGifs", () => {
    const gifs = [
      {
        id: "abc",
        url: "http://localhost:4000/cualquierimagen.jpg",
        title: "Cualquier cosa",
      },
      {
        id: "123",
        url: "http://localhost:4000/cualquierimagen.jpg",
        title: "Cualquier cosa",
      },
    ];
    useFetchGifs.mockReturnValue({ data: gifs, loading: false });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false); // Verifico que el parrafo no existe
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length); // Verifico que el parrafo no existe
  });
});
