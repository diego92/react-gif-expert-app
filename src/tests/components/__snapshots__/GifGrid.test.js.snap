// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="abc"
      key="abc"
      title="Cualquier cosa"
      url="http://localhost:4000/cualquierimagen.jpg"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Cualquier cosa"
      url="http://localhost:4000/cualquierimagen.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> debe mostrar <GifGrid /> correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <p
    className="card animate__animated animate__fadeIn"
  >
    Cargando...
  </p>
</Fragment>
`;
